lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3928,6 +3926,119 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
+  p.printKeywordOrString(classType.getElement(getIndex()).name);
+  p << "]";
+  p.printOptionalAttrDict((*this)->getAttrs(), std::array{StringRef("index")});
+  p << " : " << classType;
+}
+
+ParseResult ObjectSubfieldOp::parse(OpAsmParser &parser,
+                                    OperationState &result) {
+  auto *context = parser.getContext();
+
+  OpAsmParser::UnresolvedOperand input;
+  std::string fieldName;
+  Type inputType;
+  if (parser.parseOperand(input) || parser.parseLSquare() ||
+      parser.parseKeywordOrString(&fieldName) || parser.parseRSquare() ||
+      parser.parseOptionalAttrDict(result.attributes) || parser.parseColon() ||
+      parser.parseType(inputType))
+    return failure();
+
+  if (parser.resolveOperand(input, inputType, result.operands))
+    return failure();
+
+  auto classType = dyn_cast<ClassType>(inputType);
+  if (!classType)
+    return parser.emitError(parser.getNameLoc(),
+                            "input must be !firrtl.class type, got ")
+           << inputType;
+
+  auto index = classType.getElementIndex(fieldName);
+  if (!index)
+    return parser.emitError(parser.getNameLoc(),
+                            "unknown field " + fieldName + " in class type ")
+           << classType;
+  result.addAttribute("index",
+                      IntegerAttr::get(IntegerType::get(context, 32), *index));
+
+  SmallVector<Type> inferredReturnTypes;
+  if (failed(inferReturnTypes(context, result.location, result.operands,
+                              result.attributes.getDictionary(context),
+                              result.getRawProperties(), result.regions,
+                              inferredReturnTypes)))
+    return failure();
+  result.addTypes(inferredReturnTypes);
+
+  return success();
+}
+
+LogicalResult ObjectSubfieldOp::verify() {
+  // Verify that the field exists.
+  auto classType = getInput().getType();
+  auto index = getIndex();
+  if (classType.getNumElements() <= index)
+    return emitOpError() << "field index too large";
+
+  // Verify that the result type of this op agrees with the type of the field.
+  auto type = getType();
+  auto element = classType.getElement(index);
+  if (type != element.type)
+    return emitOpError() << "result type " << type
+                         << " does not match field type " << element.type;
+  return success();
+}
+
+LogicalResult
+ObjectSubfieldOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto classType = getInput().getType();
+  auto className = classType.getNameAttr();
+
+  // Verify that the class exists.
+  auto classOp = dyn_cast_or_null<ClassOp>(
+      symbolTable.lookupSymbolIn(circuitOp, className));
+  if (!classOp)
+    return emitOpError() << "references unknown class " << className;
+
+  // Verify that the class type agrees with the class definition.
+  return classOp.verifyType(classType, [&]() { return emitOpError(); });
```

Comment:
should the emitOpError include any extra diagnostic information? i'm not sure how the return value of the second argument to `verifyType` is used, but `return emitOpError()` seems like it should either include some error message, or just be `return failure()`. also- maybe add a failure test for this case?

Reply:
The second argument is a callback used to get an InFlightDiagnostic, and then verifyType will stream (`_<<_`) additional information to the diagnostic.

Thinking about this, the error reporting when the error is on a port is pretty bad, I think I will open a PR to clean it up. Here's a failure case that shows what that output looks like:

```mlir
firrtl.circuit "ClassTypeWrongPortDir" {
  firrtl.class @MyClass(out %str: !firrtl.string) {}
    // expected-error @below {{'firrtl.module' op port "str" has wrong direction, got in, expected out}}
  firrtl.module @ClassTypeWrongPortDir(out %port: !firrtl.class<@MyClass(in str: !firrtl.string)>) {}
}
```